[Navigation] Throw exceptions in navigation methods if in detached state

rwlbuis opened the pull request:
#### 7e1ce4ca4dde839299bd340169b10fdd6cdd6e65
<pre>
[Navigation] Throw exceptions in navigation methods if in detached state
<a href="https://bugs.webkit.org/show_bug.cgi?id=269987">https://bugs.webkit.org/show_bug.cgi?id=269987</a>

Reviewed by Darin Adler.

The back/forward/traverseTo implementation needs to check whether the document is fully active and throw an exception if not [1].
Similar rules apply to reload [2] and navigate [3].

To fix the frame/document detachment, add an IDL attribute [BypassDocumentFullyActiveCheck], which ignores the script execution context being suspended and avoids the ASSERT in DOMPromise::promise().
This behaviour matches callPromiseFunction, which simply returns the JSPromise and ignores the script execution context.

[1] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal">https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal</a> (step 5)
[2] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload">https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload</a> (step 5)
[3] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate">https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate</a> (step 7)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-already-detached-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-detach-in-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-already-detached-expected.txt:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertPromise.h:
(WebCore::JSConverter&lt;IDLPromiseIgnoringSuspension&lt;T&gt;&gt;::convert):
* Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GetBaseIDLType):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::updateCurrentEntry):
* Source/WebCore/page/Navigation.idl:

Canonical link: <a href="https://commits.webkit.org/277487@main">https://commits.webkit.org/277487@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e089ed17de0d6b0165a0f95b3a1650aa482a6b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47720 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50403 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43775 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24387 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38840 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24563 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41199 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20137 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22044 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42390 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5769 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44065 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42812 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52296 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22756 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19086 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46143 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24028 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45177 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10534 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24816 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23748 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -525,4 +531,24 @@ bool Navigation::dispatchDownloadNavigateEvent(const URL&, const String& downloa
     return false;
 }
 
+JSC::JSObject* convertDictionaryToJS(JSC::JSGlobalObject& lexicalGlobalObject, JSDOMGlobalObject& globalObject, const Navigation::Result& dictionary)
```
There are 0 other places where we needed custom convertDictionaryToJS functions in WebKit.  If a change is needed, it should be an IDL attribute.  Why doesn't the generated function work?

rwlbuis replied:
That is true. Note though that returning a promise pair is new, not well specced and likely needs some creative solution.
I thought of a few and I am not sure if they are viable:
- hardcode callPromisePairFunction to deal with returning a JS object consisting of two JS Promises (this was the initial approach). Problem is that it would be hacky to just hardcode the committed/finished names., whereas for callPromiseFunction it is just an anonymous name
- weaken the assertion in DOMPromise::promise to ignore isSuspended if the JS promise is already settled (could be checked in JSConverter<IDLPromise<T>> too). I don't know if that is acceptable outside of Navigation API though.
- what I did in my new PR. In general we could add more keywords for promise pair to the idl and react on them in the code generator, but I don't see a particular clean way. 

Of course, feel free to suggest yet another solution :)

darinadler replied:
Why can’t you just use an attribute on the type instead of a hardcoded type name? That’s all @achristensen07 is asking for, and I don’t see why it’s not practical.

rwlbuis replied:
Ok, I didn't read Alex comment carefully yesterday, at least not the " If a change is needed, it should be an IDL attribute." part. I'll experiment a bit with that.

achristensen07 commented on Source/WebCore/bindings/js/JSDOMConvertNavigationAPI.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace WebCore {
+
+template<> struct JSConverter<IDLPromise<IDLInterface<NavigationHistoryEntry>>> {
```
This seems a little better.  The day may come when we need something like this but more generic.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertPromise.h:
```diff
@@ -89,4 +91,25 @@ template<typename T> struct JSConverter<IDLPromise<T>> {
     }
 };
 
+template<typename T> struct JSConverter<IDLPromiseIgnoringSuspension<T>> {
```
We could have this derive from `JSConverter<IDLPromise<T>>` so we can share `needsState`, `needsGlobalObject`, and the `convert` template that takes a promise proxy, to make this a little less repetitive.

I also think that the `convert` function that takes `RefPtr<DOMPromise>` should probably be changed to take a `const RefPtr<DOMPromise>&` instead to avoid a bit of reference count churn.

darinadler commented on Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
```diff
@@ -64,6 +64,7 @@ template<typename T> struct IDLSequence;
 template<typename T> struct IDLFrozenArray;
 template<typename K, typename V> struct IDLRecord;
 template<typename T> struct IDLPromise;
+template<typename T> struct IDLPromiseIgnoringSuspension;
```
These forward declarations don’t need template argument names. We should remove all the `T`, `K`, and `V`.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertPromise.h:
```diff
@@ -89,4 +91,25 @@ template<typename T> struct JSConverter<IDLPromise<T>> {
+
+    static JSC::JSValue convert(JSC::JSGlobalObject&, JSDOMGlobalObject&, DOMPromise& promise)
+    {
+        return promise.guardedObject();
```
It’s a little subtle that `guardedObject` is effectively `promiseIgnoringSuspension`.

darinadler commented on Source/WebCore/bindings/scripts/IDLAttributes.json:
```diff
@@ -396,6 +396,9 @@
             "contextsAllowed": ["type"],
             "values": ["*"]
         },
+        "IgnoreSuspension": {
```
I’m not entirely sure that our terminology is ideal. There may be a better term for it that could be easier to understand in future. One example of what I mean is we could have named this `BypassDocumentActiveChecks`.

darinadler replied:
Or even better, use terminology that matches how the specification describes it. I read the specification text and now would use the name `BypassDocumentFullyActiveCheck`.

darinadler reviewed (approved):
My mild only concern is that these now-correct checks might cause some rare compatibility problems with content that worked with old Safari and WebKit; this can exist in apps and even, in some cases, obscure websites, typically not on the public web. I’m sure it will be helpful from a correctness and interoperability point of view.

rwlbuis committed:
[Navigation] Throw exceptions in navigation methods if in detached state
https://bugs.webkit.org/show_bug.cgi?id=269987

Reviewed by Darin Adler.

The back/forward/traverseTo implementation needs to check whether the document is fully active and throw an exception if not [1].
Similar rules apply to reload [2] and navigate [3].

To fix the frame/document detachment, add an IDL attribute [BypassDocumentFullyActiveCheck], which ignores the script execution context being suspended and avoids the ASSERT in DOMPromise::promise().
This behaviour matches callPromiseFunction, which simply returns the JSPromise and ignores the script execution context.

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal (step 5)
[2] https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-reload (step 5)
[3] https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate (step 7)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-already-detached-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-detach-in-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-already-detached-expected.txt:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertPromise.h:
(WebCore::JSConverter<IDLPromiseIgnoringSuspension<T>>::convert):
* Source/WebCore/bindings/js/JSDOMPromiseDeferredForward.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GetBaseIDLType):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::updateCurrentEntry):
* Source/WebCore/page/Navigation.idl:

Canonical link: https://commits.webkit.org/277487@main
